File: 106_ConstructBinaryTreeFromInorderAndPostorderTraversal.cpp
//20.29%
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if(postorder.empty())
			return NULL;
		return buildTreeHelper(inorder, 0, postorder.size() - 1, postorder, 0, postorder.size() - 1);
    }
private:
	TreeNode *buildTreeHelper(const vector<int>& inorder, int startI, int endI, const vector<int>& postorder, int startP, int endP){
		if(startI > endI)
			return NULL;
		if(startI == endI)
			return new TreeNode(inorder[startI]);
		
		TreeNode *root(new TreeNode(postorder[endP]));
		int pos;
		for(pos = startI; pos <= endI; ++pos){
			if(inorder[pos] == postorder[endP])
				break;
		}
		root -> left = buildTreeHelper(inorder, startI, pos - 1, postorder, startP, startP + pos - startI - 1);
		root -> right = buildTreeHelper(inorder, pos + 1, endI, postorder, startP + pos - startI, endP - 1);
		return root;
	}
};

//90.08%
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if(postorder.empty())
			return NULL;
		int size(postorder.size());
		TreeNode *root(new TreeNode(postorder[size - 1]));
		TreeNode *curr(root);
		stack<TreeNode *> s;
		s.push(curr);
		int pIndex(size - 2), iIndex(size - 1);
		while(pIndex >= 0){
			if(curr -> val != inorder[iIndex]){
				curr -> right = new TreeNode(postorder[pIndex--]);
				curr = curr -> right;
				s.push(curr);
			}else{
				while(!s.empty() && s.top() -> val == inorder[iIndex]){
					curr = s.top();
					s.pop();
					--iIndex;
				}
				curr -> left = new TreeNode(postorder[pIndex--]);
				curr = curr -> left;
				s.push(curr);
			}
		}
		return root;
    }
};

File: 106_ConstructBinaryTreeFromInorderAndPostorderTraversal.py
##99.21%
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        if not postorder:
            return None
        
        size = len(postorder)
        curr = root = TreeNode(postorder[size - 1])
        s = [curr]
        iIndex, pIndex = size - 1, size - 2
        while pIndex >= 0:
            if curr.val != inorder[iIndex]:
                curr.right = TreeNode(postorder[pIndex])
                curr = curr.right
                s.append(curr)
            else:
                while s and s[-1].val == inorder[iIndex]:
                    curr, iIndex = s[-1], iIndex - 1
                    s.pop()
                curr.left = TreeNode(postorder[pIndex])
                curr = curr.left
                s.append(curr)
            pIndex -= 1
        return root